deluge/ui/gtkui/mainwindow.py

```diff
@@ -109,6 +109,24 @@ def first_show(self):
             while gtk.events_pending():
                 gtk.main_iteration(False)
             self.vpaned.set_position(self.initial_vpaned_position)
+        elif deluge.common.windows_check():
+            # Realize window to create a handle for wndproc.
+            self.window.realize()
+
+        if deluge.common.windows_check():
+            # Handle win32 WM_QUERYENDSESSION message so GUI can be shutdown cleanly.
+            from win32gui import CallWindowProc, SetWindowLong
+            from win32con import GWL_WNDPROC, WM_QUERYENDSESSION
+
+            def on_wndproc(hwnd, msg, wparam, lparam):
+                # log.debug("%s, %s, %s", msg, wParam, lParam)
+                if msg == WM_QUERYENDSESSION:
+                    reactor.stop()
+                # Pass all messages on to the original WndProc.
+                return CallWindowProc(old_wndproc, hwnd, msg, wparam, lparam)
+
+            # Set WndProc to above function and store old value.
+            old_wndproc = SetWindowLong(self.window.window.handle, GWL_WNDPROC, on_wndproc)
```

Comment:
Need to fix `old_wndproc` scoping.

Additionally if dropping pywin32 need to wrap `on_wndproc` with above `WNDPROC` func.

Reply:
In what way to fix `old_wndproc` scoping?

Reply:
Needs to be a class variable when you have a callback so i'd call it `self.win32_prev_wndproc`

Also it is best to define a class variable in __init__ and assign it None.